README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# reasonabletools

<!-- badges: start -->
[![CRAN status](https://www.r-pkg.org/badges/version/reasonabletools)](https://CRAN.R-project.org/package=reasonabletools)
[![Lifecycle: maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
<!-- badges: end -->

This package provides the user with convenience functions for cleaning
and summarizing data for use in Reasonable Potential Analyses (RPAs) and
Water Quality-based Effluent Limitation (WQBEL) development. Examples of
tasks one can do with reasonabletools include constructing composite
parameter measurement (e.g., total PCBs) from congener water samples,
finding maximum observed concentrations from partially censored
datasets, and projecting maximum effluent concentrations (MECs). Methods
for projecting MECs and computing effluent limitations are based on
EPA’s Technical Support Document for Water Quality-based Toxics
Control (TSD).

## Installation

You can install the current stable version of the package from CRAN
using the following:

``` r
install.packages("reasonabletools")
```

You can install the current development version of reasonabletools from
[github](https://github.com/mattreusswig/reasonabletools) using
devtools.

``` r
# If devetools is not installed, install it.
install.packages("devtools")

# Install reasonabletools from the github repository.
devtools::install_github("mattreusswig/reasonabletools")
```

## Assumptions for CVs and Distributions

### Minimum Information Content for CV

Consistent with the TSD, reasonabletools requires a minimum level of
information content in the dataset in order to estimate the Coefficient
of Variation (CV) for a pollutant. There must be at least 10
observations and fewer than 80 percent of the dataset must be non-detect
to calculate a CV value. If either condition is not met, a CV value of
0.6 is assumed and output.

### Non-Detects Substitution

Consistent with the TSD, reasonabletools uses substitution methods when
calculating mean and standard deviation values when non-detects are
present. When non-detects are present, result values corresponding to
non-detect qual values are multiplied by 0.5 prior to computing the mean
and standard deviation.

### Distributional Assumptions

Consistent with the TSD, reasonabletools assumes a lognormal
distribution is applicable to the data when performing effluent
limitation computations.

## Usage

Functions in reasonabletools accept vectors as arguments–typically, a
vector of numeric data representing concentration point measurements or
detection limits, and a vector of character data indicating the
detected/non-detect/detected-but-not-quantified status of each numeric
value.

Combining the input vectors in a dataframe or a tibble is not necessary,
but can be useful when developing summaries for multiple pollutants at
once (see discussion below).

``` r
library(reasonabletools)

## Create an example dataset
zinc <- data.frame(qual   = c(rep("", 10), rep("<", 10)),
                   result = 1:20,
                   stringsAsFactors = FALSE)
zinc
#>    qual result
#> 1            1
#> 2            2
#> 3            3
#> 4            4
#> 5            5
#> 6            6
#> 7            7
#> 8            8
#> 9            9
#> 10          10
#> 11    <     11
#> 12    <     12
#> 13    <     13
#> 14    <     14
#> 15    <     15
#> 16    <     16
#> 17    <     17
#> 18    <     18
#> 19    <     19
#> 20    <     20

## Find observed MEC
find_mec(qual = zinc$qual, result = zinc$result)
#>   qual result
#> 1          10
```

### Non-Detects, Detected Values, and DNQs

Functions in reasonabletools can distinguish between detected and
non-detect values. Non-detect values are indicated in the qualifier
vector with pre-specified character string. By default, this is a “\<”,
“nd”, or “ND” character string. Any string in the non-detect qualifier
vector which does not match these indicators is treated as a detected
value. However, the non-detect flags can be changed using the “nd”
argument (e.g., you could change it to “My ND Indicator”–see example
below). In function output, “\<” is always used to indicate a censored
value and an empty string "" is used to indicate a detected value.

``` r
## Create an example dataset with new custom non-detect indicator
zinc <- data.frame(qual   = c(rep("My ND Indicator", 20)),
                   result = 1:20,
                   stringsAsFactors = FALSE)
zinc
#>               qual result
#> 1  My ND Indicator      1
#> 2  My ND Indicator      2
#> 3  My ND Indicator      3
#> 4  My ND Indicator      4
#> 5  My ND Indicator      5
#> 6  My ND Indicator      6
#> 7  My ND Indicator      7
#> 8  My ND Indicator      8
#> 9  My ND Indicator      9
#> 10 My ND Indicator     10
#> 11 My ND Indicator     11
#> 12 My ND Indicator     12
#> 13 My ND Indicator     13
#> 14 My ND Indicator     14
#> 15 My ND Indicator     15
#> 16 My ND Indicator     16
#> 17 My ND Indicator     17
#> 18 My ND Indicator     18
#> 19 My ND Indicator     19
#> 20 My ND Indicator     20

## Find observed MEC
find_mec(qual = zinc$qual, result = zinc$result, 
         nd = c("My ND Indicator"))
#>   qual result
#> 1    <      1
```

When cleaning data for summary, recall that R is a case-sensitive
language and capitalization matters in non-detect flags. The default
censored flags in reasonabletools are “\<”, “nd”, and “ND”. This means
an observation in your data marked “nd” will be treated as a non-detect,
but an observation marked “nD” (notice the change in capitalization on
the second letter) will be treated as a detected value.

All elements in the results vector should have a numeric value–do not
include NA values. For non-detect values, the method detection limit
should be entered in the results column. NA values in the results column
are ignored and will result in incorrect results.

Functions in reasonabletools do not have specialized behavior for an
observation which is detected-but-not-quantified (DNQ or j-flag). The
user must decide when preprocessing the data whether these values should
be treated as non-detect values or as detected values. The observation
should be marked as non-detect or detected in the qualifier vector, and
the reporting limit or estimated concentration placed in result vector.

## Working with Multiple Pollutants

Real-life datasets used in NPDES permit development will typically
include many pollutants of concern which need to be summarized. This is
a split-apply-combine task which is cumbersome to address in Excel and
other spreadsheet-based programs for anything but simple summary
functions (e.g., computing averages or maximums without conditionality).

In base R (or using dplyr/tidyverse methods) this task becomes
straightforward. Here are some strategies from base R and dplyr for
summarizing effluent data with reasonabletools.

We will use the following toy dataset in these examples:

``` r
set.seed(10010)
data <- data.frame(pollutant = sort(rep(c("zinc", "arsenic", "copper",
                                          "lead", "nickel", "pcbs"), 5)),
                   qualifier = sample(c("<", ""), size = 30, 
                                      replace = TRUE, prob = c(0.8, 0.2)),
                   result = sample(seq(0.1, 0.5, 0.1), 30, replace = TRUE),
                   sampling.date = rep(seq.Date(from = as.Date("2012-03-10"), 
                                                to = as.Date("2012-03-10") + 365 * 4,
                                                by = 365), 6),
                   stringsAsFactors = FALSE)

data
#>    pollutant qualifier result sampling.date
#> 1    arsenic         <    0.1    2012-03-10
#> 2    arsenic              0.3    2013-03-10
#> 3    arsenic         <    0.4    2014-03-10
#> 4    arsenic         <    0.1    2015-03-10
#> 5    arsenic         <    0.3    2016-03-09
#> 6     copper         <    0.5    2012-03-10
#> 7     copper         <    0.4    2013-03-10
#> 8     copper         <    0.5    2014-03-10
#> 9     copper         <    0.5    2015-03-10
#> 10    copper         <    0.2    2016-03-09
#> 11      lead         <    0.1    2012-03-10
#> 12      lead         <    0.1    2013-03-10
#> 13      lead         <    0.2    2014-03-10
#> 14      lead         <    0.4    2015-03-10
#> 15      lead         <    0.4    2016-03-09
#> 16    nickel              0.5    2012-03-10
#> 17    nickel              0.2    2013-03-10
#> 18    nickel         <    0.4    2014-03-10
#> 19    nickel         <    0.1    2015-03-10
#> 20    nickel         <    0.3    2016-03-09
#> 21      pcbs         <    0.2    2012-03-10
#> 22      pcbs              0.4    2013-03-10
#> 23      pcbs         <    0.5    2014-03-10
#> 24      pcbs         <    0.2    2015-03-10
#> 25      pcbs         <    0.4    2016-03-09
#> 26      zinc              0.4    2012-03-10
#> 27      zinc         <    0.3    2013-03-10
#> 28      zinc         <    0.5    2014-03-10
#> 29      zinc              0.2    2015-03-10
#> 30      zinc         <    0.3    2016-03-09
```

### Base R

Use the split family of functions to create a list of dataframes grouped
by pollutant name, use lapply to apply the function to each element of
the list, and do.call + rbind to recombine the output.

``` r

data_split <- split.data.frame(x = data, f = data$pollutant)

## Note we need to use an anonymous function in lapply for functions with
## multiple non-default input arguments, like find_mec.
data_apply <- lapply(X = data_split, 
                     FUN = function(x) find_mec(x$qualifier, x$result))

combined_summary <- do.call(rbind, data_apply)

combined_summary
#>         qual result
#> arsenic         0.3
#> copper     <    0.2
#> lead       <    0.1
#> nickel          0.5
#> pcbs            0.4
#> zinc            0.4
```

Or, we can combine the apply and combine steps into a single line by
nesting the functions. Unfortunately, we can’t do this in fewer than two
steps.

``` r
data_split <- split.data.frame(x = data, f = data$pollutant)

combined_summary <- do.call(
       rbind, 
       lapply(X = data_split, 
              FUN = function(x) find_mec(x$qualifier, x$result))
  )

combined_summary
#>         qual result
#> arsenic         0.3
#> copper     <    0.2
#> lead       <    0.1
#> nickel          0.5
#> pcbs            0.4
#> zinc            0.4
```

### dplyr (v1.0 or later)

Using dplyr is often a more intuitive way to implement
split-apply-combine for many people. Using dplyr, we will use
dplyr::group\_by and wrap the reasonabletools summary function in
dplyr::summarise and it will automatically output all columns grouped by
pollutant. (Note that old versions of dplyr::summarise (pre-1.0) are not
set up to output two columns from a single summary function–see the next
section for pre-1.0 dplyr strategies.)

``` r
library(dplyr)

data %>%
  group_by(pollutant) %>%
  summarise(find_mec(qualifier, result))
#> # A tibble: 6 x 3
#>   pollutant qual  result
#>   <chr>     <chr>  <dbl>
#> 1 arsenic   ""       0.3
#> 2 copper    "<"      0.2
#> 3 lead      "<"      0.1
#> 4 nickel    ""       0.5
#> 5 pcbs      ""       0.4
#> 6 zinc      ""       0.4
```

### dplyr (Pre-v1.0)

Older versions of dplyr were not set up to handle multiple column output
in the summarise functions. There were two main workarounds for this–one
built into reasonabletools and one using dplyr itself.

A number of data summarising functions in reasonabletools that output
two column values, like find\_mec or project\_mec, include an option for
automatically concatenating the two output vectors into a single
character vector. The “simple\_output” argument will produce a single
character vector when set to TRUE, and will produce two vectors when
left at its default FALSE value.

``` r
library(dplyr)

## simple_output = TRUE turns output into single character vector.
data %>% 
  group_by(pollutant) %>% 
  summarise(MEC = find_mec(qualifier, result, simple_output = TRUE))
#> # A tibble: 6 x 2
#>   pollutant MEC  
#>   <chr>     <chr>
#> 1 arsenic   0.3  
#> 2 copper    <0.2 
#> 3 lead      <0.1 
#> 4 nickel    0.5  
#> 5 pcbs      0.4  
#> 6 zinc      0.4
```

Note this setting is also useful when constructing summary tables for
inclusion in permit documents where one wants to present the qualifier
and value in the same table cell.

If one wants both output vectors retained in the output, use the
dplyr::do function as a wrapper around the desired reasonabletools
function. Use a dot prior to the $ operator and column name (e.g.,
.$qualifier) to indicate to dplyr::do it should look for the inputs in
the piped dataframe.

``` r

data %>% 
  group_by(pollutant) %>% 
  do(find_mec(.$qualifier, .$result)) 
#> # A tibble: 6 x 3
#> # Groups:   pollutant [6]
#>   pollutant qual  result
#>   <chr>     <chr>  <dbl>
#> 1 arsenic   ""       0.3
#> 2 copper    "<"      0.2
#> 3 lead      "<"      0.1
#> 4 nickel    ""       0.5
#> 5 pcbs      ""       0.4
#> 6 zinc      ""       0.4
```


news.md
# reasonabletools 0.1

- Release of reasonabletools.

R/summary_functions.R
#' @title Find the adjusted coefficient of variation
#' @description Calculates adjusted coefficient of variation (CV) according to methods described in EPA's Technical Support Document for Water Quality-based Toxics Control.
#'
#' @param qual  A character vector containing non-detect indicator strings, e.g., "<" or "ND". The strings used to indicate censored status can be edited in the "nd" argument.
#' @param result A numeric vector of concentration measurements.
#' @param nd A list indicating all the censoring flags included in the dataset. Defaults to "<", "nd", and "ND".
#' @param nd_adjustment Adjustment factor for non-detect values. Non-detect values (as indicated in qual vector) are multiplied by nd_adjustment factor; i.e. result * nd_adjustment. Typically, method detection limits or reporting limits are used as result values for non-detects.
#'
#' @return A numeric coefficient of variation (CV) value
#' @export
#'
#' @examples 
#' # CV for all detected values 
#' cen_result <- rep("", 10)
#' result     <- c(1:10) 
#' cv_adj(cen_result, result)
#' 
#' # CV for all non-detected values
#' cen_result <- rep("<", 10)
#' cv_adj(cen_result, result)
#' 
#' # CV for fewer than 10 measurements
#' cen_result <- rep("", 5)
#' result <-     c(1:5)
#' cv_adj(cen_result, result)
#' 
#' # Change the default substitution value
#' cen_result <- c(rep("<", 5), rep("", 15))
#' result     <- c(101:120)
#' cv_adj(cen_result, result)   # Use default 0.5 multiplier
#' # Use 1.0 multiplier (equivalent to using MDL)
#' cv_adj(cen_result, result, nd_adjustment = 1.0)  
#' # Use 0.0 multiplier (equivalent to zero substitution)
#' cv_adj(cen_result, result, nd_adjustment = 0)  
cv_adj <- function(qual, result, 
                   nd = c("<", "nd", "ND"), nd_adjustment = 0.5) {
  
  
  ## Error message is given if the the 'result' argument is not numeric.
  result_class <- class(result)
  if (!is.numeric(result)) stop("result arugment must be numeric. The result
                               argument entered is a ", result_class, " type.",
                                        call. = FALSE)
  

  ## Error message is given if the the 'result' and 'qual' arguments are of
  ## unequal lengths.
  if (length(result) != length(qual)) stop("The qual and result arguments must be equal in length. Qual is ", length(qual), " elements long, and result is ", length(result), ".", call. = FALSE)
  
  
  ## Main body of function
  
  # Helper function which creates an anti-union operator (opposite of %in%)
  "%not_in%" <- function(x, y) { !(x %in% y) }
  
  ## Index which observations are detects and which are nondetects
  detects <- which(qual %not_in% nd)             ## gives index of detects
  nondetects <- which(qual %in% nd)              ## gives index of nondetects
  
  
  ## The CV is conditional on data quality (i.e., n and proportion of censored
  ## values)
  n <- length(result)                           ## gives number of observations
  censored_prop <- length(nondetects) / n       ## gives fraction of non-detects
  
  
  ## Compute CV conditionally. If low data quality, CV == 0.6
  if ( (n < 10) | (censored_prop >= 0.8) ) { cv <- 0.6 
  
  
  ## If acceptable data quality, CV is the ratio of the sd to mean.
  ## Adjust censored values to the desired substitution values using
  ## the nd_adjustment argument.
  } else {
    
    ## Issue a warning when NA's are present which is unexpected behavior.
    if (anyNA(result)) {
      na_index <- which(is.na(result))
      na_index <- paste(na_index, collapse = ", ")
      warning(paste0("Your result argument contains NA values at index loacation(s) ", na_index, " which have been removed in this calculation. Your coefficient of variation estimate may contain errors."))
    } 
    
    ## Compute CV
    std_deviation <- stats::sd( c(result[detects], 
                           nd_adjustment * result[nondetects]), 
                         na.rm = TRUE)
    
    average       <- mean( c(result[detects], 
                             nd_adjustment * result[nondetects]),
                           na.rm = TRUE)
    
    cv <- std_deviation / average  
    
  }
  
  return(cv)        ## Exit
  
}


#' @title Find the Maximum Observed Effluent Concentration (MEC)
#' @description Find the MEC (no projection) from the observed dataset using methods described in EPA's Technical Support Document for Water Quality-based Toxics Control.
#'
#' @param qual  A character vector containing non-detect indicator strings, e.g., "<" or "ND". The strings used to indicate censored status can be edited in the "nd" argument.
#' @param result A numeric vector of concentration measurements.
#' @param nd A list indicating all the censoring flags included in the dataset. Defaults to "<", "nd", and "ND". 
#' @param simple_output Logical value. If TRUE, the output columns are concatenated into a single character string (e.g., "<0.2") which can be useful for constructing summary tables. 
#'
#' @return Dataframe with a qualifier column (character) and a MEC column (numeric).
#' @export
#'
#' @examples
#' # Find observed MEC
#' cen_result <- c(rep("", 10), rep("<", 10))
#' result     <- 1:20
#' find_mec(cen_result, result)
#' 
#' cen_result <- rep("<", 20)
#' find_mec(cen_result, result)
#' 
#' cen_result <- rep("", 20)
#' find_mec(cen_result, result)
#' 
#' # Demonstrate simplified output
#' find_mec(cen_result, result, simple_output = TRUE)
#' 
#' # Define a set of custom non-detect flags
#' cen_result <- c(rep("non-detect", 5), rep("<", 10), rep("mdl", 5))
#' find_mec(cen_result, result, nd = c("non-detect", "<", "mdl"))
#' 
find_mec <- function(qual, result, 
                     nd = c("<", "nd", "ND"), simple_output = FALSE) {
  
  ## Error message is given if the the 'result' argument is not numeric.
  result_class <- class(result)
  
  if (!is.null(result)) {
    if (!is.numeric(result)) stop("The result arugment must be numeric. The result argument entered is a ", result_class, " type.",
                                          call. = FALSE)
  }
  
  ## Error message is given if the the 'result' and 'qual' arguments are of
  ## unequal lengths.
  if (length(result) != length(qual)) stop("The qual and result arguments must be equal in length. Qual is ", length(qual), " elements long, and result is ", length(result), ".", call. = FALSE)
  
  ## consider adding a helper function that looks for character strings not
  ## included in the provided "nd" list within the qual vector, and warn
  ## the user they are present.
  
  
  ## Main body of function
  
  # Helper function which creates an anti-union operator (opposite of %in%)
  "%not_in%" <- function(x, y) { !(x %in% y) }
  
  # Create indices for detected and non-detected results
  detects <- which(qual %not_in% nd)    ## gives index of detects
  nondetects <- which(qual %in% nd)     ## gives index of nondetects
  
  # How many observations are there
  n <- length(result)
  n_nd <- length(nondetects)
  
  all_nd <- if (n_nd == n) TRUE else FALSE
  
  if (all_nd == TRUE) {
    
    mec <- data.frame(qual = "<", result = min(result, na.rm = TRUE), 
                      stringsAsFactors = FALSE)

  }  
  
  if (all_nd == FALSE) {
    
    mec <- data.frame(qual = "", result = max(result[detects], 
                                              na.rm = TRUE), 
                      stringsAsFactors = FALSE)
    
  }
  
  if (isTRUE(simple_output)) mec <- paste0(mec[[1]], mec[[2]])
  
  return(mec)
  
}


#' @title Find the projected Maximum Observed Effluent Concentration (MEC)
#' @description Find the MEC projected from a lognormal distribution using methods described in EPA's Technical Support Document for Water Quality-based Toxics Control.
#'
#' @param qual  A character vector containing non-detect indicator strings, e.g., "<" or "ND". The strings used to indicate censored status can be edited in the "nd" argument.
#' @param result A numeric vector of concentration measurements.
#' @param nd A list indicating all the censoring flags included in the dataset. Defaults to "<", "nd", and "ND". 
#' @param percentile Location on lognormal distribution to estimate the projected MEC.
#' @param conf_level Confidence level of projected estimate.
#' @param nd_adjustment Adjustment factor for non-dectect values. Non-detect values (as indicated in qual vector) are multiplied by nd_adjustment factor; i.e. result * nd_adjustment. Typically, method detection limits or reporting limits are used as result values for non-detects.
#' @param simple_output Logical value. If TRUE, the output columns are concatenated into a single character string (e.g., "<0.2") which can be useful for constructing summary tables.
#'
#' @return  Dataframe with a qualifier column (character) and a MEC column (numeric).
#' @export
#'
#' @examples
#' # Find observed MEC
#' cen_result <- c(rep("", 10), rep("<", 10))
#' result     <- 1:20
#' project_mec(cen_result, result)
#' 
#' # Demonstrate simplified output
#' cen_result <- rep("<", 20)
#' project_mec(cen_result, result, simple_output = TRUE)
#' 
#' # Define a set of custom non-detect flags
#' cen_result <- c(rep("non-detect", 5), rep("<", 10), rep("mdl", 5))
#' project_mec(cen_result, result, nd = c("non-detect", "<", "mdl"))
#' 
#' # Change the substitution multiplier used for non-detect values
#' cen_result <- c(rep("<", 5), rep("", 15))
#' result     <- c(101:120)
#' project_mec(cen_result, result)   # Use default 0.5 multiplier
#' 
#' # Use 1.0 multiplier (equivalent to using MDL)
#' project_mec(cen_result, result, nd_adjustment = 1.0)
#' 
#' # Use 0.0 multiplier (equivalent to zero substitution)  
#' project_mec(cen_result, result, nd_adjustment = 0)  
#' 
project_mec <- function(qual, result, 
                        nd = c("<", "nd", "ND"),
                        percentile = 0.95, conf_level = 0.99,
                        nd_adjustment = 0.5, simple_output = FALSE) {
  
  
  ## Error message is given if the the 'result' argument is not numeric.
  result_class <- class(result)
  
  if (!is.numeric(result)) stop("The result arugment must be numeric. The result argument entered is a ", result_class, " type.",
                                        call. = FALSE)
  
  
  ## Error message is given if the qual and result args do not have the same
  ## number of elements.
  n_qual = length(qual)
  n_rslt = length(result)
  
  if (n_qual != n_rslt) stop("The qual and result vectors must have an equal number of elements. The qual argument has ", n_qual, " elements and the result argument has ", n_rslt, " elements.")
  
  
  ## Find the indices of all the nondetect and detect observations
  detects    <- which(!(qual %in% nd))  ## gives index of detects
  nondetects <- which(qual %in% nd)  ## gives index of nondetects
  
  
  ## If all observations are non-dectect, then return the 
  ## minimum detection limit.
  all_nd <- if (length(nondetects) == length(qual)) TRUE else FALSE
  
  if (all_nd == TRUE) {
    
    mec = data.frame(qual = "<", 
                     result = min(result, na.rm = TRUE),
                     stringsAsFactors = FALSE)
  }   
  
  
  ## If at least one detected value then calculate a projected 
  ## MEC based on n and CV.
  if (all_nd == FALSE) {
    ## Compute the sample max effluent concentration.
    df <- find_mec(qual = qual, result = result, 
                   nd = nd, simple_output = FALSE)
    
    max_obs_result <- df[[2]]
    
    
    ## Calculate the total number of observations, and the location--as a
    ## percentile (p_n) of the sample mec.
    n_obs <- length(result)
    p_n <- (1 - conf_level)^(1/n_obs)
    
    
    ## Uses cv_adj() from this script to compute coeff of variation. 
    cv <- cv_adj(qual = qual, result = result, 
                 nd = nd, nd_adjustment = nd_adjustment)
    
    
    ## Compute sigma for a lognormal distribution, according to the TSD
    sigma <- sqrt(log(1 + cv^2))  ## log() is the natural log (i.e. ln) in R
    
    
    ## Compute the mec multiplier, according to the TSD method
    c_target   <- exp(stats::qnorm(percentile) * sigma - 0.5 * sigma^2)
    c_observed <- exp(stats::qnorm(p_n) * sigma - 0.5 * sigma^2)
    
    
    ## Compute the MEC
    mec <- data.frame(qual = "", 
                      result = max_obs_result * (c_target / c_observed), 
                      stringsAsFactors = FALSE)
  }
  
  ## If simple_output == TRUE, then convert the 2 column mec dataframe to
  ## a single character value
  if (isTRUE(simple_output)) mec <- paste0(mec[[1]], mec[[2]])
  
  return(mec)                                            ## EXIT
  
}



#' @title Compute Wasteload Allocation
#' @description Compute the wasteload allocation (WLA) used for effluent limit calculation.
#'
#' @param criteria Limiting water quality criterion. Must be in same units as background argument.
#' @param background Background pollutant concentration. Must be in same units as criteria argument.
#' @param Qrsw Upstream limiting/design receiving water flowrate. Must be in same units as Qeff argument. Flow arguments may be entered in ratio form.
#' @param Qeff Effluent limiting/design flowrate. Must be in same units as Qrsw argument.  Flow arguments may be entered in ratio form.
#'
#' @return WLA as numeric value in same units as criteria and background.
#' @export
#'
#' @examples
#' # WLA for pollutant with 2 ug/L acute criteria and upstream receiving water concentration 
#' # of 0.1 ug/L. The critical flows are 3 MGD (1Q10) and 0.5 MGD (max daily flow).
#' calc_WLA(2, 0.1, 3, 0.5) 
#' 
#' # When using dilution credits, put Qrsw and Qeff in terms of the dilution ratio (D).
#' D = 7   # Assume a jurisdiction that uses D = (Qrsw + Qeff) / Qeff
#' Qeff = 1 # Equal to 1 since its the denominator of the ratio, or you can use the critical flow
#' Qrsw = D - 1 # Same as the expression (D * Qeff) - Qeff
#' calc_WLA(2, 0.1, Qrsw = D - 1, Qeff = 1)
calc_WLA <- function(criteria, background, Qrsw, Qeff) {
  
  ## For dilution ratios, set Qeff = 1 and Qrsw to appropriate value
  
  Q <- Qrsw + Qeff
  
  WLA <- (Q * criteria - Qrsw * background) / Qeff
  
  return(WLA)                                           ## EXIT
   
}



#' @title Compute MDL
#' @description Compute Maximum Daily Effluent Limitation (MDL) according to methods in EPA's Technical Support Document for Water Quality-based Toxics Control.
#'
#' @param WLAa Numeric. Acute wasteload allocation (WLA). All WLA units should be identical.
#' @param WLAc Numeric. Chronic wasteload allocation (WLA). All WLA units should be identical.
#' @param WLAhh Numeric. Human health wasteload allocation (WLA). All WLA units should be identical.
#' @param cv Numeric. Coefficient of variation (CV) of effluent data. See cv_adj function.
#' @param n_samples Numeric. Number of sample observations.
#' @param prob_LTA Numeric (fraction). Allowable exceedance probability of the WLA used to estimate long-term average (LTA).
#' @param percentile_MDL Numeric (fraction). Lognormal distribution location for MDL. 
#' @param percentile_AML Numeric (fraction). Lognormal distribution location for AML.
#'
#' @return Numeric value in same units as the WLAs.
#' @export
#'
#' @examples 
#' calc_MDL(WLAa=4, WLAc=1, WLAhh=10, cv=0.6)
#' 
calc_MDL <- function(WLAa, WLAc, WLAhh, cv, 
                    n_samples = 4, prob_LTA = 0.99, 
                    percentile_MDL = 0.99, percentile_AML = 0.95) {
  
  ##TK add error function if arguments are non-numeric
  ##TK add error if arguments are not all of same length
  ##TK add error for negative values
  
  ## Step 1: Find the WLA for each criteria
  ## Given in function arguments
  
  ## Step 2: Compute a LTA for each 
  ### Compute the log variance (sigma_2) of the effuent data
  sigma_2  <- log(1 + cv^2)
  sigma4_2 <- log(1 + 0.25 * cv^2)
  
  ### Compute the acute and chronic multipliers
  acute_mult <- exp(0.5 * sigma_2 - stats::qnorm(prob_LTA) * sqrt(sigma_2))
  chron_mult <- exp(0.5 * sigma4_2 - stats::qnorm(prob_LTA) * sqrt(sigma4_2))
  
  ### Compute the LTAs
  LTAa <- WLAa * acute_mult
  LTAc <- WLAc * chron_mult
  
  ## Step 3: Find controlling LTA
  LTA <- min(LTAa, LTAc, na.rm = TRUE)
  
  ## Step 4: Compute the MDL for aquatic life
  sigman_2 <- log(1 + (1 / n_samples) * cv^2)
  MDL_mult <- exp(stats::qnorm(percentile_MDL) * sqrt(sigma_2) - 0.5 * sigma_2)
  AML_mult <- exp(stats::qnorm(percentile_AML) * sqrt(sigman_2) - 0.5 * sigma_2)
  MDLaq    <- LTA * MDL_mult
  
  ## Step 5: Compute the MDL for human health
  AMLhh <- WLAhh
  MDL_to_AML_ratio <- MDL_mult / AML_mult
  MDLhh <- AMLhh * MDL_to_AML_ratio
  
  ## Step 6: Select the controlling (minimum) MDL
  MDL <- min(MDLaq, MDLhh, na.rm = TRUE)
  
  
  return(MDL)                        ## EXIT
}


#' @title Compute AML
#' @description Compute average monthly effluent limitation (AML) according to methods in EPA's Technical Support Document for Water Quality-based Toxics Control.
#'
#' @param WLAa Numeric. Acute wasteload allocation (WLA). All WLA units should be identical.
#' @param WLAc Numeric. Chronic wasteload allocation (WLA). All WLA units should be identical.
#' @param WLAhh Numeric. Human health wasteload allocation (WLA). All WLA units should be identical.
#' @param cv Numeric. Coefficient of variation (CV) of effluent data. See cv_adj function.
#' @param n_samples Numeric. Number of sample observations.
#' @param prob_LTA Numeric (fraction). Allowable exceedance probability of the WLA used to estimate long-term average (LTA). 
#' @param percentile_AML Numeric (fraction). Lognormal distribution location for AML.
#'
#' @return Numeric value in same units as the WLAs.
#' @export
#'
#' @examples 
#' calc_AML(WLAa=4, WLAc=1, WLAhh=10, cv=0.6)
#' 
calc_AML <- function(WLAa, WLAc, WLAhh, cv, 
                    n_samples = 4, prob_LTA = 0.99, percentile_AML = 0.95) {
  
  ##TK add error function if arguments are non-numeric
  ##TK add error if arguments are not all of same length
  ##TK add error for negative values
  
  ## Step 1: Find the WLA for each criteria
  ## Given in function arguments
  
  ## Step 2: Compute a LTA for each 
  ### Compute the log variance (sigma_2) of the effuent data
  sigma_2  <- log(1 + cv^2)
  sigma4_2 <- log(1 + 0.25 * cv^2)
  
  ### Compute the acute and chronic multipliers
  M_acute <- exp(0.5 * sigma_2 - stats::qnorm(prob_LTA) * sqrt(sigma_2))
  M_chron <- exp(0.5 * sigma4_2 - stats::qnorm(prob_LTA) * sqrt(sigma4_2))
  
  ### Compute the LTAs
  LTAa <- WLAa * M_acute
  LTAc <- WLAc * M_chron
  
  ## Step 3: Find controlling LTA
  LTA <- min(LTAa, LTAc, na.rm = TRUE)
  
  ## Step 4: Compute the aquatic life AML
  ### The log variance is a function of the expected number of samples taken
  ### each month. Typically assume 4 for most toxics unless sampling is more
  ### frequent (e.g., daily = 30).
  sigman_2 <- log(1 + (1 / n_samples) * cv^2)
  AMLaq <- LTA * exp(stats::qnorm(percentile_AML) * sqrt(sigman_2) - 0.5 * sigma_2)
  
  ## Step 5: Compute the human health AML
  AMLhh <- WLAhh
  
  ## Step 6: Return the controlling (minimum) AML
  AML <- min(AMLaq, AMLhh, na.rm = TRUE)
  
  return(AML)                        ## EXIT
}
